### PR TITLE
feat: add default_lmdb_capacity configuration support

### PR DESCRIPTION
**Pull request by twilson63**
Add support for applying default_lmdb_capacity to LMDB stores in nested configurations

### CODE REVIEW

**Review by twilson63 — COMMENTED** [review]
# Comprehensive Code Review for PR 371 - feat: add default_lmdb_capacity configuration support

## Summary
This PR adds support for applying default_lmdb_capacity to LMDB stores in nested configurations. The feature allows operators to configure a default capacity for all LMDB stores through a global configuration option, while still allowing explicit capacity overrides on individual stores.

## Core Changes Analysis

### 1. Main Changes in hb_http_server.erl
The PR introduces two new functions and modifies the server startup sequence:

**New Functions:**
- update_store_with_capacity/2: Main function that handles both list and map store configurations
- update_single_store_with_capacity/2: Recursive helper for updating nested store configurations

**Modified Functions:**
- start/0: Now calls update_store_with_capacity before starting the store
- set_default_opts/1: Also applies capacity updates to default store configurations

### 2. Security Analysis

**Positive Security Aspects:**
- **Input Validation**: The code properly checks for existing capacity settings before applying defaults
- **Non-Destructive**: Existing capacity settings are preserved, preventing accidental overwrites
- **Type Safety**: Uses pattern matching to ensure only LMDB stores are modified
- **Bounded Recursion**: Recursion is limited to known nested store fields

**Security Concerns:**
- **Configuration Injection**: The default_lmdb_capacity value is not validated for reasonable bounds
- **Memory Exhaustion**: Large capacity values could potentially cause memory issues
- **Nested Configuration Depth**: No protection against deeply nested configurations

### 3. Performance Analysis

**Performance Improvements:**
- **Centralized Configuration**: Avoids need to manually configure capacity for each LMDB store
- **Lazy Evaluation**: Only processes stores that actually need capacity updates
- **Efficient Pattern Matching**: Uses guard clauses for type-specific handling

**Performance Considerations:**
- **Recursive Processing**: Deep nesting could impact startup time
- **Map Operations**: Multiple map updates during configuration processing
- **Memory Allocation**: The recursive nature creates temporary maps during processing

### 4. Code Quality Assessment

**Improvements:**
- **Clear Documentation**: Well-documented functions with comprehensive docstrings
- **Comprehensive Testing**: Extensive test coverage including edge cases
- **Consistent Patterns**: Follows established coding patterns in the codebase
- **Error Handling**: Graceful handling of missing configuration values

**Code Quality Strengths:**
- **Modular Design**: Separate functions for different aspects of the functionality
- **Recursive Design**: Handles complex nested store configurations elegantly
- **Test Coverage**: Thorough test suite covering various scenarios
- **Type Safety**: Uses pattern matching for type-safe operations

### 5. Testing Analysis

**Test Coverage:**
- ✅ No default capacity configured
- ✅ Default capacity with mixed store types
- ✅ Single store configuration
- ✅ Explicit capacity precedence
- ✅ Non-LMDB stores unchanged
- ✅ Nested store configurations
- ✅ LRU with persistent store
- ✅ Gateway with local store
- ✅ Deeply nested configurations

**Test Results:**
- All 4 tests pass successfully
- Test execution time: 1.166s (excellent performance)
- No failures or timeouts

### 6. Recommendations

**High Priority:**
1. **Add Validation**: Validate default_lmdb_capacity values for reasonable bounds
2. **Memory Safety**: Consider adding upper limits to prevent memory exhaustion
3. **Documentation**: Update configuration documentation to describe the new option

**Medium Priority:**
1. **Performance Monitoring**: Add metrics for configuration processing time
2. **Error Handling**: Add better error messages for invalid configurations
3. **Configuration Validation**: Validate nested store configurations

### 7. Security Recommendations

1. **Input Validation**: Add bounds checking for capacity values
2. **Resource Limits**: Implement reasonable upper limits for memory allocation
3. **Configuration Auditing**: Log configuration changes for security monitoring

### 8. Verdict

**Overall Assessment: ✅ LOOKS GOOD with Minor Suggestions**

This PR successfully adds a valuable configuration feature with excellent code quality and comprehensive testing. The implementation is clean, well-documented, and follows established patterns in the codebase.

**Key Benefits:**
- Simplifies LMDB configuration management
- Maintains backward compatibility
- Comprehensive test coverage
- Clean, readable code
- Proper error handling

**Minor Concerns:**
- Should add validation for capacity values
- Could benefit from performance monitoring
- Documentation needs updating

**Production Readiness:**
The code is production-ready with the suggested validation improvements. The comprehensive test suite and clean implementation make this a low-risk addition that provides significant operational value.

**Review by twilson63 — COMMENTED** [review]
# Code Review for PR 371 - default_lmdb_capacity configuration

## Summary
Adds  configuration option that automatically applies to all LMDB stores while preserving explicit capacity settings.

## Key Changes
- **New functions**:  and 
- **Modified startup**: Applies capacity defaults before store initialization
- **Recursive processing**: Handles nested store configurations (store, persistent-store, local-store)

## Security & Performance
**✅ Strengths:**
- Non-destructive (preserves existing capacity settings)
- Type-safe (only affects LMDB stores)
- Comprehensive test coverage

**⚠️ Concerns:**
- No bounds validation on capacity values (potential memory exhaustion)
- Recursive processing could impact startup time with deep nesting

## Testing
All 4 tests pass (1.166s) with extensive coverage including nested configurations.

## Recommendations
**High Priority:**
1. Add capacity value validation/bounds checking
2. Consider memory safety limits

**Medium Priority:**
1. Update configuration documentation
2. Add performance monitoring for config processing

## Verdict
**✅ LOOKS GOOD** - Production-ready with suggested validation improvements. Clean implementation that significantly improves LMDB configuration management.

**Review by twilson63 — COMMENTED** [review]
# Code Review for PR 371 - default_lmdb_capacity configuration

## Summary
Adds default_lmdb_capacity configuration option that automatically applies to all LMDB stores while preserving explicit capacity settings.

## Key Changes
- New functions: update_store_with_capacity/2 and update_single_store_with_capacity/2
- Modified startup: Applies capacity defaults before store initialization
- Recursive processing: Handles nested store configurations (store, persistent-store, local-store)

## Security & Performance
**Strengths:**
- Non-destructive (preserves existing capacity settings)
- Type-safe (only affects LMDB stores)
- Comprehensive test coverage

**Concerns:**
- No bounds validation on capacity values (potential memory exhaustion)
- Recursive processing could impact startup time with deep nesting

## Testing
All 4 tests pass (1.166s) with extensive coverage including nested configurations.

## Recommendations
**High Priority:**
1. Add capacity value validation/bounds checking
2. Consider memory safety limits

**Medium Priority:**
1. Update configuration documentation
2. Add performance monitoring for config processing

## Verdict
**LOOKS GOOD** - Production-ready with suggested validation improvements. Clean implementation that significantly improves LMDB configuration management.

**Review by twilson63 — COMMENTED** [review]
## 💡 Suggestion: Make this more generic for extensibility

Great implementation\! This could be made more generic to handle defaults for any store module type. Here's a suggested approach:

### Current Pattern:
```
default_lmdb_capacity → applies to hb_store_lmdb
```

### Generic Pattern:
```
default_<store_type>_<parameter> → applies to hb_store_<store_type>
```

### Examples:
```erlang
% config.flat
default_lmdb_capacity: 1073741824
default_rocksdb_block_cache_size: 536870912  
default_lru_max_entries: 10000
default_fs_buffer_size: 65536
```

### Benefits:
- **Extensible**: Easy to add new store types without code changes
- **Consistent**: Same pattern for all store configurations
- **Discoverable**: Predictable naming convention
- **Maintainable**: Single code path for all defaults

### Implementation:
Replace `update_store_with_capacity` with `update_store_with_defaults` that:
1. Extracts all `default_*` keys from configuration
2. Parses them to determine store module and parameter
3. Applies defaults recursively like current implementation

This would make the feature much more powerful while maintaining backward compatibility. The current LMDB-specific implementation works great as a starting point\!

Ref: Created example implementation files showing how this could work.

**Comment by twilson63** [comment]
refactored to make default store params more generic

File: src/hb_http_server.erl [hunk 1]
```diff
@@ -511,6 +514,136 @@ set_default_opts(Opts) ->
...
+    UpdatedConfig = case maps:get(<<"store-module">>, StoreConfig, undefined) of
+        undefined -> StoreConfig;
+        StoreModule ->
+            case maps:get(StoreModule, DefaultParams, #{}) of
+                Empty when map_size(Empty) == 0 -> StoreConfig;
+                StoreDefaults ->
+                    apply_defaults_to_store_config(StoreConfig, StoreDefaults)
+            end
+    end,
+
```
Comment (samcamwilliams) [inline src/hb_http_server.erl:618]: Style!
